src/main/java/com/limechain/network/protocol/transactions/TransactionsEngine.java

```diff
@@ -0,0 +1,142 @@
+            return;
+        }
+        connectionManager.addTransactionsStream(stream);
+        log.log(Level.INFO, "Received transactions handshake from " + peerId);
+        //writeTransactionsMessage(stream, peerId);
+    }
+
+    private void handleResponderStreamMessage(byte[] message, Stream stream) {
+        System.out.println("On Responder");
+        PeerId peerId = stream.remotePeerId();
+        boolean connectedToPeer = connectionManager.isTransactionsConnected(peerId);
+
+        if (!connectedToPeer && !isHandshake(message)) {
+            log.log(Level.WARNING, "No handshake for transactions message from Peer " + peerId);
+            stream.close();
+            return;
+        }
+
+        if (isHandshake(message)) {
+            handleHandshake(peerId, stream);
+        } else {
+            handleTransactionMessage(message, peerId);
+        }
+    }
+
+    private void handleHandshake(PeerId peerId, Stream stream) {
+        if (connectionManager.isTransactionsConnected(peerId)) {
+            log.log(Level.INFO, "Received existing transactions handshake from " + peerId);
+            stream.close();
+        } else {
+            connectionManager.addTransactionsStream(stream);
+            log.log(Level.INFO, "Received transactions handshake from " + peerId);
+        }
+        writeHandshakeToStream(stream, peerId);
+    }
+
+    private void handleTransactionMessage(byte[] message, PeerId peerId) {
+        ScaleCodecReader reader = new ScaleCodecReader(message);
+        byte[][] transactions = reader.read(new TransactionsReader());
+        log.log(Level.INFO, "Received " + transactions.length + " transactions from Peer "
+                + peerId);
+        //TODO Add transactions to data
+        }
+
+    /**
+     * Send our Transactions handshake on a given <b>initiator</b> stream.
+     *
+     * @param stream <b>initiator</b> stream to write the message to
+     * @param peerId peer to send to
+     */
+    public void writeHandshakeToStream(Stream stream, PeerId peerId) {
+        byte[] handshake = new byte[]{};
+
+        log.log(Level.INFO, "Sending transactions handshake to " + peerId);
+        stream.writeAndFlush(handshake);
+    }
+
+    /**
+     * Send our Transactions message from {@link SyncedState} on a given <b>responder</b> stream.
+     *
+     * @param stream <b>responder</b> stream to write the message to
+     * @param peerId peer to send to
+     */
+    public void writeTransactionsMessage(Stream stream, PeerId peerId) {
+        ByteArrayOutputStream buf = new ByteArrayOutputStream();
+        try (ScaleCodecWriter writer = new ScaleCodecWriter(buf)) {
+            writer.write(new TransactionsWriter(), new byte[][]{new byte[]{}, new byte[]{}});
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        log.log(Level.INFO, "Sending transaction message to peer " + peerId);
+        //TODO Send our transaction message
```

Comment:
Is this todo still relevant?

Reply:
Yes. Transactions queue algorithms must be implemented otherwise we do not know if the peer has already received this transaction from us.